### DISCUSSION TITLE
In-game panel sprites update

### DISCUSSION THREAD

**Discussion by mefistotelis** [swfans/syndwarsfx#238] [General]
As an offshoot from work on in-game panel, I started looking into fonts. For now, the more detailed ones. There are two such fonts:
* the menu uses 23pt font
* within the game no large font is loaded, but there are sprites created from 20pt font

Both fonts are now in [https://github.com/swfans/swars-gfx] repo. While I didn't modified the larger one too much, for the smaller one - I generated it by rescaling 23pt to 20pt, and then fixing it manually to make it more symmetric and geometrically more straight.

So the question now is, whether that will improve the in-game fonts or not. There are only 3 uses of this font:

![text_uplink](https://github.com/user-attachments/assets/ad417132-e412-4e07-a06d-56080eafaf5b)
![text_paused](https://github.com/user-attachments/assets/33c020e9-0bad-436d-823b-cd2b037305b0)
![text_detail](https://github.com/user-attachments/assets/01821df3-791f-41c7-a6e2-9f1dd869605b)

And here are the originals for comparison:
![text_uplink-orig](https://github.com/user-attachments/assets/f07183b8-f7ee-4f61-b330-97bc341b403d)
![text_paused-orig](https://github.com/user-attachments/assets/b0226e4e-75cc-4345-a5f6-c1e132d11545)
![text_detail-orig](https://github.com/user-attachments/assets/0678bcd6-94a9-4bd8-a926-bafe70eedb15)

Based on these, let me know if any letters should be improved. Accepting feedback both as text and as pull request to swars-gfx.

**Comment by mefistotelis** [comment]
Broadened the subject - from fonts, to all panel sprites.

The `swars-gfx` package now contains all `pop*-*.dat/tab` and `panel*-*.dat/tab`. The `pop`s contain sprites for normal game panels, the `panel`s are older HUD, without per-agent weapon list (this currently can only be enabled by editing game code; will likely chnage at some point).

What I'm getting to is that all panel sprites are now taken from `swars-gfx` repo, and most of the sprites are automatically generated using command line graphics tool. This means they differ from originals, because they're no longer hand-edited, but also because I "invented" the transformations to make them. This was required from various reasons; anyway, while generating them I might have missed something not looking right - so please test and report if all weapons look correctly, whether nothing randomly shifts one pixel, and so on.

Generated sprites are for all panel styles (style change = F9), and both resolutions (F8).

**Reply by Moburma** [reply]
Had a play with them, look pretty good! Can hardly tell the difference with the default and solid blue designs. Green one is less good, for me the highlight on the selected weapons is too blurry/indistinct compared to the original, needs to be a bit sharper. Also I got a bug with the default blue design, the unselected weapons all appear with a black rectangle (this is just downloading the latest package from github actions and the latest executable).

Green original vs new:
![](https://i.ibb.co/vZWyJxH/dos4gw-216.png)
![](https://i.ibb.co/n8fbYg0/Desktop-Screenshot-2025-01-08-10-43-04-29.png)

Hmm, it might just be the difference in resolution in this one, I will try again with same res and compare..

Blue panel bug:

![](https://i.ibb.co/S0CvBhY/Desktop-Screenshot-2025-01-08-10-47-08-57.png)

**Reply by mefistotelis** [reply]
> bug with the default blue design, the unselected weapons all appear with a black rectangle

My local builds are fine - so it must be the difference in ImageMagick version. Will investigate (though GitHub moves to newer Ubuntu this month, with newer IM installed, so this issue would disappear by itself).

**Reply by mefistotelis** [reply]
What a mess.. looks like the rule is "after using morphology filter on an image, you cannot save PNG. Instead, pipe the result directly to another imagemagick process and save the PNG there".

So, it's clearly a bug in older IM. I assumed some functions just changed the way they work, and took me hours to finally figure out that was not the case.

**Reply by mefistotelis** [reply]
For green crazygas - here's the original vs mine.
Note that the highlight reaches a bit further, but is toned down a bit.
Also, I'm selecting the pixels inside the shape and decreasing their intensity even more.

Do you still think we should tone down the glow? I mean we can do it, that would make the shape more readable. Just that such change wouldn't bring it closer to original, but the opposite. But I'm definitely not against changes. Back in the days I never used that panel - but if I remember correctly it wasn't because the glow, but because the hidden guns were too low contrast.

Original first, generated second.
![green-crazygas-original](https://github.com/user-attachments/assets/5b94966d-d3c9-4213-a774-6f309c60fbc7)
![weapon_outline_crazygas foup lgt](https://github.com/user-attachments/assets/6439d966-7cdb-42b0-9726-2c635414d90e)

EDIT:
I also now see that I stopped deleting one line of the symbol to make it look like original.. should get back to deleting it.

**Reply by Moburma** [reply]
> What a mess.. looks like the rule is "after using morphology filter on an image, you cannot save PNG. Instead, pipe the result directly to another imagemagick process and save the PNG there".
> 
> So, it's clearly a bug in older IM. I assumed some functions just changed the way they work, and took me hours to finally figure out that was not the case.

Actually, do you think now would be a good idea to do a new "full" release of swars using the old graphics package etc? Then it's in a known good state for these changes to follow. Plus it's a new year, so people would be interested in progress over last year etc.

**Reply by mefistotelis** [reply]
>  do a new "full" release

I recently messed the input on low res panels, and it may take some time to straighten up. So not right now; but we should look for a window for a release.

I will focus on fixing the low res rather than continuing with preparations to make the panels scalable. And yeah, will try not to mess with the sprite numbers so that the original files still work.

**Reply by mefistotelis** [reply]
> do a new "full" release

Now it probably a best moment for the release; do you have a preferred gfx and sfx package build to use?

**Reply by Moburma** [reply]
Probably make sense to use the current pointer graphics package, and perhaps do a new sound one with the dome sound and mech included?

**Reply by mefistotelis** [reply]
> make sense to use the current pointer graphics package,

if we want a package with cursors, but without fonts and panels, then this might be better:
https://github.com/swfans/swars-gfx/actions/runs/12078335600

> do a new sound one with the dome sound and mech included?

ok, so sfx to the latest commit.

I will prepare the sfx/gfx releases and update the main repo. Then you can create the release. Is that ok?

**Reply by Moburma** [reply]
Sure

**Reply by mefistotelis** [reply]
Packages have new releases, references in main `swars` are updated. Your turn.

**Reply by Moburma** [reply]
I've made a draft in releases, have a look. I concentrated more on end user features, so a lot of the tons of back end stuff is not included. Obviously feel free to add more stuff that you think is relevant. I didn't do bugs this time as a lot of the features are really bug fixes to original game code.
Called out the fact people need to get get used to getting things from the other repositories, or use the installer.

**Reply by mefistotelis** [reply]
Updated to mind Linux. No other remarks. Looks good to me.

**Reply by geist22** [reply]
@Moburma are you using the default/english version of MAP065.MAD (moon) in the levels package? It might make sense to use the korean version instead where the mech can path out of its enclosure. It doesn't seem to have any downsides to it (e.g. when not using the mech), does it?

**Reply by Moburma** [reply]
Hmm, I was thinking about this the other day, I agree. It's actually using the original map currently, but I think the Korean one is better. The only downside is it makes the level easier if you're not using my modified level with the boss, as you can ride the mech yourself instead.

Then again a) most people won't know you can do this b) they can always exercise self-control and not use it! Even getting near there is pretty hard, so I think it's ok.

**Reply by mefistotelis** [reply]
I think we can agree that a map with the path open is better. For changes in level file, we can discuss separately.

The question is, does that Korean version introduce any additional changes, ones we wouldn't want to take.
(I remember looking into it in the past, but I no longer remember the conclusion)

**Reply by Moburma** [reply]
I've used it myself extensively and not noticed anything else different/negative. That said there are quite a few changes comparing the binaries.

**Reply by mefistotelis** [reply]
I remember testing walking (with teleport) on the central structure. So I've probably seen some differences in pathfinding around it. But the area is unreachable anyway, so that doesn't matter that much.

**Comment by geist22** [comment]
A bit late reply, but I'm afraid I'm not a fan of the changed fonts shown in the first post, due to things like uneven corners and diagonals, and some characters looking too different from the originals.

Do you have to originals somewhere in unpacked form (png, bmp, etc.)? I couldn't find them in the gfx repo, only the edited ones. I'm thinking about trying some editing myself to see if it turns out useable, though I'm a bit busy at the moment.

**Reply by mefistotelis** [reply]
> Do you have to originals somewhere in unpacked form (png, bmp, etc.)

Here's a full pack of originals:
https://mega.nz/file/dR9lySKb#WwpvAj_k2LjAOdNnAnKwmO0XEacuOe5Bqp9MkbRvLhY

The panels are `POP*`, and they include the words as single bitmaps.
In `swars-gfx` repo, you will not find corresponding whole bitmaps, because they're merged together from single characters. The single characters for 640x480 mode sprites are in `swars-gfx/font0_sharp_cult-20-white`.

**Comment by mefistotelis** [comment]
We would use a bit smaller sprites than the sets I recently created. Ie. `pointers-96` (meaning 75% size of `pointers-128`), and `weapons-outline-27` (75% of `weapons-outline-36`). And `panel_*-96` as well.

The rules: they need to be a transparent PNG without partial transparency (so every pixel either transparent or not, no soft edges).

Anyone interested in giving it a try?

**Reply by geist22** [reply]
I could take a look, but it would take a while (probably a few weeks).

For scaling, you could either source everything from the originals, or create the biggest version first and then scale everything else down from that. Which ones are the originals, `POINTERS`, `W-ICONS` and `PANEL*-*`?

I'm currently working on a complete replacement of `font0_sharp_cult-20-white` and about halfway done with it.

By the way, is there a reason you embedded the color profiles in the files you made? I don't think this is necessary, and it also increases the file size.

**Comment by mefistotelis** [comment]
> I could take a look, but it would take a while (probably a few weeks).

Well, just write here when you start to make sure noone else will work on the same folder.

> For scaling, you could either source everything from the originals, or create the biggest version first and then scale everything else down from that.

Most of the files I mentioned already have higher res versions (that's why I mentioned 75% of bigger size rather than 150% of smaller size).

> Which ones are the originals, POINTERS, W-ICONS and PANEL*-*?

In most folders you will find `.txt` files used to create specific `.dat`/`.tab`, the names of these clearly indicate the final file name.
Weapons folder lack the txt files, as the weapons are not used to create any file directly; instead, they're recolored and effects are added to generate panel files.

So currently we're talking about the files: `panel*`, `pop*` and `pointr*`. The last one of these was renamed from `pointers`.

> By the way, is there a reason you embedded the color profiles in the files you made?

Which file has an embedded profile? While GFX software stores references to a profile, I tried to not keep them embedded.
But really, I don't care even about the reference. At the final conversion, the files are treated as linear RGB.

**Comment by mefistotelis** [comment]
Here's how the larger sprites look currently; not fully supported for several reasons.
But what this shot made me realize is that these panels are just too large, meaning their height is too large.
So I'm leaning towards using lower res panels for this resolution, stretching them horizontally in a smart way. "Smart" meaning only the bar area for each agent will be stretched, weapon and agent number will remain as is. This will completely break compatibility with original panel sprite files, that's why I worked on the way of generating the sprite files.

![synII196](https://github.com/user-attachments/assets/799bb725-398d-4422-98a8-5e487d3eb4c6)

**Comment by geist22** [comment]
Those are the images taken from the menu screens? It might be difficult to scale them down, probably requires some improvisation and compromise on the details. It could be more preferrable to scale up the less detailed ingame icons for that size.

>Which file has an embedded profile?

So far every file I've opened from the gfx repo, for example the entire `font0_sharp_cult-23-white` (and 20, though I intend to replace those later) folder. From other folders I've only taken a look at single files here and there, though they had it as well. You might want to re-check everyting you've made in the gfx repo.

I'll probably also save my files as 8-bit png instead of 24-bit, since they use less than 256 colors. It's not much, but I like to optimize where I can.

**Comment by mefistotelis** [comment]
> Those are the images taken from the menu screens?

Those are images I created based on higher resolution in-menu equip screen shapes and lower resolution in-game shapes.

> It might be difficult to scale them down, probably requires some improvisation and compromise on the details.

Yes. If it was as simple as selecting a scale down filter in a graphics tool, it would be already done.

> It could be more preferrable to scale up the less detailed ingame icons for that size.

That's up to the preference of the person doing the scaling. I generally prefer starting with the more complex shape, as this makes it easier for me to decide what needs simplification and what can stay as is.

> I'll probably also save my files as 8-bit png instead of 24-bit, since they use less than 256 colors.

For the upload, please keep it as 24-bit. I don't want to adjust my scripts for handling various formats of input files.

**Comment by geist22** [comment]
I have finished my version of `font0_sharp_cult-20-white`. You can find it here:
https://github.com/geist22/swars-gfx/tree/main/font0_sharp_cult-20-white

I've intended for it to completely replace the current version. I tried to make it resemble the source (the 23 pixel version) as closely as possible, but due to the constraints of having fewer pixels to work with I was forced to take some liberties. Generally I'm content with how it turned out, though some characters might need minor adjustments depending on how the'll look in action.

I don't know what some of the characters are and how they're supposed to look (`char_u00aa` and `char_u00b0`), and why they're higher placed.
Are `char_u005b` and `char_u005d` supposed to be square brackets? I don't know why they are taller than the other characters, but since it was like that in the source, I've kept them 2 pixels taller, too.

I also made an alternative `@` that is less pixelated, though I don't know how well it fits the theme.

What do you think?

**Reply by mefistotelis** [reply]
Will check it. For what each character means - I used unicode markings of characters in file name. Searching for ie. "unicode 00b0" should give you details about each character.

**Reply by mefistotelis** [reply]
Checked. I see no problem with accepting that to the repo. for the "@" symbol - replace the original with your version. I don't think the game uses that symbol anywhere, so it's for future changes only. But still, good to have as it's generally an often used char.

**Comment by mefistotelis** [comment]
The detail=3 (so for resolution (3+1)*320 = 1280) panels are now ready. At least all the issues which were bugging me are now fixed. So, please let me know if you think anything needs improving (using latest build from `swars` and overwriting sprite files with latest `swars-gfx` build).

The panels should now also be scaled correctly to any resolution.

Pause screen is still not scaled currently.

**Reply by geist22** [reply]
I don't have a build environment set up for `swars-gfx`, could you do a release?

**Reply by mefistotelis** [reply]
You can download recent builds from "Actions" tab.

**Reply by geist22** [reply]
Oh right, didn't think about that.

Looks good so far in my 1280x960 window 👍

Though one thing that's off is the border highlight on hover:
![border_highlight](https://github.com/user-attachments/assets/dae6e284-1ce6-425d-8ae8-dddf567dabaf)
Left is original, right is port. If you look at the top and bottom lines, you see the highlight is more like a gradient and loses its color towards the middle, while the original is one color for most of the line.

**Reply by mefistotelis** [reply]
> Though one thing that's off is the border highlight on hover:

These are generated using `-sparse-color  Shepards` command in ImageMagick.
Here's how the command to do that looks in the build log within "Actions":
https://github.com/swfans/swars-gfx/actions/runs/12978328331/job/36192769806#step:11:2774
And here's the documentation of said command:
https://legacy.imagemagick.org/Usage/canvas/#shepards

In the log, you see stuff like `%[fx:w*.1]` instead of a specific coordinate - this is to make thew points scale to image size - in this specific case this means: the coordinate is image width X 0.1, so a coordinate at 10% of the image.

Here's the simplified command to just generate the fill gradient:

```
convert -size 178x52 xc:white -channel RGB    -sparse-color  Shepards '%[fx:w*.1],%[fx:h*.1] rgb(43%,72%,88%)  %[fx:w*.5],%[fx:h*.5] rgb(38%,27%,88%) %[fx:w*.9],%[fx:h*.1] rgb(42%,30%,99%)  %[fx:w*.9],%[fx:h*.9] rgb(43%,72%,88%)'         +channel "panel_weapon_frame_hlight_filling.png"
```

**Reply by mefistotelis** [reply]
A bit more to make it easier to understand:
```
%[fx:w*.1],%[fx:h*.1] rgb(43%,72%,88%) - turquoise at top left
%[fx:w*.5],%[fx:h*.5] rgb(38%,27%,88%) - dark blue at center
%[fx:w*.9],%[fx:h*.1] rgb(42%,30%,99%) - lighter blue at top right
%[fx:w*.9],%[fx:h*.9] rgb(43%,72%,88%) - turquoise at bottom right
```

If you want a single gradient at top then I guess you'll have to drag the 2nd point lower, and maybe make its color lighter.

**Reply by geist22** [reply]
I've spent some time trying to make it look more like the original, but I don't know enough about how ImageMagick works. The `shepards` command spreads the color too far even beyond other dots, so to get rid of undesired gradients I'd do something like this:

`-sparse-color Voronoi '%[fx:w*.1],%[fx:h*.3] rgb(43%,73%,89%) %[fx:w*.2],%[fx:h*.6] rgb(43%,47%,97%) %[fx:w*.8],%[fx:h*.4] rgb(43%,47%,97%) %[fx:w*.9],%[fx:h*.7] rgb(43%,73%,89%)' \`

Which would need the edges blurred for a smaller gradient (I don't know how to do that). But the second panel that extends up to the left towards the medikit button then doesn't work at all with it so I gave up for now.

Also, I just noticed this:
(left is original, right is port)
![agent](https://github.com/user-attachments/assets/ca69c1b5-4a47-4f3b-b326-5fffb4e0ef4b)
The agent's right arm is detached from the body and you can see the background through the gap. Also visible from behind.

**Reply by mefistotelis** [reply]
> I gave up for now.

Yeah, this can be updated at any time.

> The agent's right arm is detached from the body

That should be a new "issue". I don't think it's related to the panels. Or is it?

**Reply by geist22** [reply]
> That should be a new "issue". I don't think it's related to the panels. Or is it?

Yes, I will open an issue. I wasn't sure if this was caused by the program itself or by the new gfx package, but testing old versions with the current gfx package display properly, so it must be from the executable.

**Comment by mefistotelis** [comment]
How it looks now:
![swars-port-panel-scale](https://github.com/user-attachments/assets/bb994c8c-d9b7-4276-8caf-c5f05a4d720d)

I don't see anything I missed in regard to the scaling.. though it's likely there are things I didn't noticed.
So, what is now required are the sprites in different resolutions. Like the detail=2 ones I mentioned before. But also detail=7 would be nice (double of detail=3 - for width 2560 and up). And also detail=5 (for width 1920), as that's a popular resolution.

In general, each detail point adds 320 to width and 240 to height (though that depends on proportions). And when files with additional detail levels are ready, the game will automatically use them.

When it can't find the detail it wants, it will open lower detail one and stretch it in a smart way.

**Comment by bobf5** [comment]
Out of curiosity, is there a way to force lower detail HUD sprites at high resolutions? I remember playing a build about a month ago that had the lower detail sprites with smart scaling, and looked surprisingly good.

**Reply by mefistotelis** [reply]
If you delete high res sprite files, the game will load lower res ones. You can even force it to load 320x200 (detail=0) panels.

I haven't decided yet which sprites will stay at 1280x720. We don't yet have sprites for width=960, and those would probably fit best.

EDIT:
Forgot about - if deleting sprites, you have to leave `pop2-3.*` intact - that one is used to reserve memory for these files, so it's necessary. It stores the green panel.
